sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamCalcRelType.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.plan.RelOptRule;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.plan.RelTraitSet;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.RelNode;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexCall;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexDynamicParam;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexFieldAccess;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexLiteral;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexLocalRef;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexNode;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexProgram;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.sql.SqlOperator;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.sql.validate.SqlUserDefinedFunction;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.tools.RelBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link org.apache.beam.sdk.extensions.sql.impl.CalcRelSplitter.RelType} for {@link BeamCalcRel}.
+ */
+class BeamCalcRelType extends CalcRelSplitter.RelType {
+  private static final Logger LOG = LoggerFactory.getLogger(BeamCalcRelType.class);
+
+  BeamCalcRelType(String name) {
+    super(name);
+  }
+
+  @Override
+  protected boolean canImplement(RexFieldAccess field) {
+    // Don't implement freestanding field accesses. Only implement field accesses that are operands
+    // to UDF calls.
+    return false;
+  }
+
+  @Override
+  protected boolean canImplement(RexLiteral literal) {
+    // Don't implement freestanding literals. Only implement literals that are operands to UDF
+    // calls.
+    return false;
+  }
+
+  @Override
+  protected boolean canImplement(RexDynamicParam param) {
+    return false;
+  }
+
+  @Override
+  protected boolean canImplement(RexCall call) {
+    final SqlOperator operator = call.getOperator();
+
+    CallImplementor implementor = RexImpTable.INSTANCE.get(operator);
+    if (implementor == null) {
+      // Reject methods with no implementation
+      return false;
+    }
+
+    if (operator instanceof SqlUserDefinedFunction) {
+      SqlUserDefinedFunction udf = (SqlUserDefinedFunction) call.op;
+      if (udf.function instanceof ZetaSqlScalarFunctionImpl) {
+        ZetaSqlScalarFunctionImpl scalarFunction = (ZetaSqlScalarFunctionImpl) udf.function;
+        if (!scalarFunction.functionGroup.equals(
+            BeamZetaSqlCatalog.USER_DEFINED_JAVA_SCALAR_FUNCTIONS)) {
+          // Reject ZetaSQL Builtin Scalar Functions
+          return false;
+        }
+        for (RexNode operand : call.getOperands()) {
+          if (operand instanceof RexLocalRef) {
+            if (!(BeamJavaUdfCalcRule.udfSupportsLiteralType(operand.getType())
```

Comment:
Just looking at the type without looking at what the `RexLocalRef` is referencing seems like a problem here. We know that some types work as inputs but not literals, this doesn't appear to enforce that distinction. (This goes back to my concern above.)

Reply:
> We know that some types work as inputs but not literals

Which ones?

I'm not sure we need to validate operand types here at all, since we do that during query parsing now (#14464).

Reply:
BigInt is one example of a (historically) problematic type: https://github.com/apache/beam/pull/14241#discussion_r594746821

Reply:
But BigInt should be supported by both now: https://github.com/apache/beam/blob/8e6695643b69da1ab83c9aad16a913cf6e442823/sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamJavaUdfCalcRule.java#L127